velox/exec/prefixsort/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,356 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/common/base/SimdUtil.h"
+
+namespace facebook::velox::exec::prefixsort {
+
+namespace detail {
+/// Provides LegacyRandomAccessIterator but not ValueSwappable semantic
+/// iterator for PrefixSort`s sort algorithm, see:
+/// https://en.cppreference.com/w/cpp/algorithm/sort
+/// https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+/// https://en.cppreference.com/w/cpp/named_req/ValueSwappable
+/// The difference is operator*(), the LegacyInputIterator return "reference,
+/// convertible to value_type", see:
+/// https://en.cppreference.com/w/cpp/named_req/InputIterator
+/// In contrast, PrefixSortIterator returns char* point to the prefix data
+/// (not a "value_type" but a memory buffer store normalized keys) for memcmp
+/// and memcpy operators, means not ValueSwappable.
+/// This is also the reason we can not use std::sort directly, and implement
+/// quickSort in PrefixSortContext.
+struct PrefixSortIterator {
+  PrefixSortIterator(char* prefix, const uint64_t entrySize)
+      : prefix_(prefix), entrySize_(entrySize) {}
+
+  PrefixSortIterator(const PrefixSortIterator& other)
+      : prefix_(other.prefix_), entrySize_(other.entrySize_) {}
+
+  // Inline functions should be place in include file to avoid link error.
+  FOLLY_ALWAYS_INLINE char* operator*() const {
+    return prefix_;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator& operator++() {
+    prefix_ += entrySize_;
+    return *this;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator& operator--() {
+    prefix_ -= entrySize_;
+    return *this;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator++(int) {
+    auto tmp = *this;
+    prefix_ += entrySize_;
+    return tmp;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator--(int) {
+    auto tmp = *this;
+    prefix_ -= entrySize_;
+    return tmp;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator+(const uint64_t i) const {
+    auto result = *this;
+    result.prefix_ += i * entrySize_;
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator-(const uint64_t i) const {
+    PrefixSortIterator result = *this;
+    result.prefix_ -= i * entrySize_;
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator& operator=(
+      const PrefixSortIterator& other) {
+    prefix_ = other.prefix_;
+    return *this;
+  }
+
+  FOLLY_ALWAYS_INLINE uint64_t
+  operator-(const PrefixSortIterator& other) const {
+    return (prefix_ - other.prefix_) / other.entrySize_;
```

Comment:
ditto
Also shall we check (prefix_ - other.prefix_) is a multiple of entrySize_?